doc/design/block.md

```diff
@@ -0,0 +1,210 @@
+# Design Doc: Use Block in RNNOp, While Op, IfElseOp
+
+In C++ and Java programming language, a block is a lexical structure of source code which is grouped as one line of code.
+
+RNNOp looks like the loop structure in programming languages.
+And similarly, WhileOp and IfElseOp are like loop and conditions respectively.
+So we want to verify if we should have a class Block in PaddlePaddle that works like a pair of curly braces in the loop and condition structures of programming languages.
+
+Blocks do not only group source code, but also narrow the lexical scope of variables so that they do not conflict with variables having the same name used elsewhere in a program.
+
+In Paddle, we need a similar concept called Block to support following scenes:
+
+- define a PaddlePaddle program by writing blocks of codes, which includes the definitions of variables and operators.
+  - `RNNOp`, `SwitchOp`, `WhileOp` and `IfElseOp`, etc, need Block to help to define sub-block. 
+- help to execute multiple operators, blocks should group operators and runs like a single operator.
+
+## How to use Block
+In `RNNOp`, `SwitchOp`, `WhileOp` and `IfElseOp`, a sub-block should be used to help to define a sub-block.
+
+Let's start from how a RNNOp is described using Block:
+
+```python
+v = some_op()
+m_boot = some_op()
+
+W = pd.Variable(shape=[20, 20])
+U = pd.Varable(shape=[20, 20])
+
+rnn = create_RNNOp()
+with rnn.stepnet() as net:
+  # declare the input variables that need to be segmented into steps
+  x = net.add_input(v)
+  # declare rnn's memory (state)
+  h = net.add_memory(init=m_boot)
+  
+  fc_out = pd.matmul(W, x)
+  hidden_out = pd.matmul(U, h.pre_memory())
```

Comment:
`h.pre_memory()` is a bad name because `h` itself is a `memory`, it is no need be `xxx_memory` again. Maybe `h` enough is OK.